### DISCUSSION TITLE
How to use visualization scripts?

### DISCUSSION THREAD

**Discussion by jaccus** [nightroman/Invoke-Build#220] [Q&A] [answered]
Do I get this right that the only way to use visualization tools (such as Show-BuildGraph.ps1 script) is via cloning the repo / copy-pasting the script into consumer code-base?

These scripts are not distributed together with IB PS module (unlike e.g. Show-TaskHelp or Resolve-MSBuild).
These scripts are not embedded into the IB dotnet tool (AFAIK).

These scripts can not also be called directly from PS (after mentioned copy-paste approach), because they require IB module to be available in the session.

So far my best idea for incorporating this into a consumer code-base is to:
1) copy-paste it somewhere into the code-base (e.g. under `.\tools\IB\Show-BuildGraph.ps1`)
2) add IB task to my root app.build.ps1 which calls the copy-pasted script, e.g. `task Graph { & "$PSScriptRoot\tools\IB\Show-BuildGraph.ps1" }`
3) call `Graph` task on my root IB script as usual

If these scripts were at least available in the IB module itself, we'd be able to simplify this process quite a bit.

**Comment by nightroman** [comment, marked as answer]
The distributed package/module contains the reasonable minimum of tools. There are pros and cons of this approach and people who like this and not, of course.

The README says "see PSGallery" and provides the list of some scripts distributed via PSGallery

> Extra scripts, see PSGallery and the repository:

[Invoke-Build.ArgumentCompleters](https://www.powershellgallery.com/packages/Invoke-Build.ArgumentCompleters) - completers for v5 native, TabExpansion2.ps1
[Invoke-TaskFromVSCode](https://www.powershellgallery.com/packages/Invoke-TaskFromVSCode) - invokes a task from a build script opened in VSCode
[Show-BuildGraph](https://www.powershellgallery.com/packages/Show-BuildGraph) - shows task graph by Graphviz Viz.js or dot
[New-VSCodeTask](https://www.powershellgallery.com/packages/New-VSCodeTask) - generates VSCode tasks bound to build script tasks
[Invoke-TaskFromISE](https://www.powershellgallery.com/packages/Invoke-TaskFromISE) - invokes a task from a script opened in ISE

These scripts may be installed or saved locally by these commands:

```powershell
Install-Script Show-BuildGraph
Save-Script Show-BuildGraph -Path ...
```

The idea is that end users decide if they need these tools or not. If they do then they have a choice to install scripts once and use for all IB scripts now and then. Also, these scripts are usually developed and may be updated on their own, separately from IB.
